### PR TITLE
Refactoring: Move sound sources into src/sound folder

### PR DESCRIPTION
**Pull request by ann0see**
**Short description of changes**

Moves the sound.cpp / sound.h / sound.mm files to src/sound/

CHANGELOG: Refactoring: move sound API files into src/sound folder. Contributors will now find the sound.cpp/sound.h/sound.mm in the src/sound folder for consistency

**Context: Fixes an issue?**

Related to: https://github.com/jamulussoftware/jamulus/issues/2573

@pgScorpio, I assume you should also have a look at this PR. 

**Does this change need documentation? What needs to be documented and how?**
Yes

**Status of this Pull Request**

Waiting on second review

**What is missing until this pull request can be merged?**

Second review

## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want (works on Linux and Windows. Since the CI is clean it should be ok)
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

### CODE REVIEW

**Comment by pgScorpio** [comment]
@ann0see 
Yes, I will keep an eye on this PR too...

And if we indeed are going to change the src tree we maybe should have a look if it wouldn't be useful to split up some more   at the same time... i.e. Common, Client, Server, Protocol,....

And would it be possible to get the same structure in the editors (Qt Creator, VS, XCode) too? That would be a great improvement too.

**Review by hoffie — COMMENTED** [review]
Didn't have a deeper look yet, just some thoughts. As said elsewhere, I like this idea.

- [ ] Did you check translations? It looks like each .ts file contains references to the source path, so I guess they must/should be updated to match the new location to avoid all translations showing up as new? I haven't checked though if this is really the case. Maybe Qt tooling is intelligent enough to update it itself, but it should be checked nevertheless.
- [x] Various checks or scripts contain explicit references to our code directories. As everything is moved to the existing src/, this should be fine already, but I'm mentioning it nevertheless.
- [ ] I didn't think of the fact that windows/ etc. are not gone automatically, as they contain build scripts, platform-specific resources (icons, installers). Some (at least mac/) continues to ship source files. Would be nice to get rid of that as well, but maybe in a follow-up PR...
- [x] I think it would be logical to move the windows/ASIOSDK2 extraction target. Either, it should be moved along with the code which is using it or it could be moved to libs/ where all other third-party stuff lives. As this is closely related to the move of the ASIO sound implementation, I think this should be done in this PR as well. It will likely affect docs and build scripts though.

**Comment by ann0see** [comment]
Hmm. Moving ASIO around involves some bigger changes, I think?

I'd like to have it in libs.

**Comment by ann0see** [comment]
@hoffie just updated the .ts files

**Comment by ann0see** [comment]
> I didn't think of the fact that windows/ etc. are not gone automatically, as they contain build scripts, platform-specific resources (icons, installers). Some (at least mac/) continues to ship source files. Would be nice to get rid of that as well, but maybe in a follow-up PR...

Yes. But this needs a bit of restructuring other places too...

File: Jamulus.pro
Comment (ann0see) [inline Jamulus.pro:1123]: @hoffie I think you should check if this is ok here.

**Comment by pgScorpio** [comment]
> Hmm. Moving ASIO around involves some bigger changes, I think?
> 
> I'd like to have it in libs.

I agree that that seems more consistent. oboe is also in libs.
Though I don't see yet why that would involve bigger changes. Changing base folder is changing base folder, whether it is src/sound/asio/ or src/libs/ doesn't make that much of a difference or do I mis something here ?

**Comment by ann0see** [comment]
No. I did that already in this PR - and it seems to build fine.

File: Jamulus.pro
Comment (pgScorpio) [inline Jamulus.pro:198]: Why still repeating the same lines again and again?
Whenever something changes in the source files needed for a specific sound implementation I would like to have to change that at just one location.
So who not using separate defines for the sound folders as I suggested in [this issue](https://github.com/jamulussoftware/jamulus/issues/2573) ?

So using
```
HEADERS+=$$HEADERS_JACK
SOURCES+=$$SOURCES_JACK
```
etc...

File: COMPILING.md
Comment (pgScorpio) [inline COMPILING.md:93]: Isn't it easier if we make the folder libs/ASIOSDK2 already exist, so we can just say:

Download the [ASIOSDK](https://www.steinberg.net/asiosdk), open the top level folder in the .zip file and copy the contents into `[\path\to\jamulus\source]\libs\ASIOSDK2` so that, e.g., the folder `[\path\to\jamulus\source]\libs\ASIOSDK2\common` exists

File: Jamulus.pro
Comment (pgScorpio) [inline Jamulus.pro:128]: I think it is better to check for libs/ASIOSDK2/common 

That's conform compiling.md "so that, e.g., the folder `[\path\to\jamulus\source]\libs\ASIOSDK2\common` exists" )

File: Jamulus.pro
Comment (pgScorpio) [reply Jamulus.pro]: I think line 1122 could change to `CLANG_FORMAT_SOURCES = $$find(CLANG_FORMAT_SOURCES, ^\(src)/)`
since all sources are under src now ?

Comment (ann0see) [reply Jamulus.pro]: No. There's still some cpp code in these folders(androiddebug.cpp) for example

Comment (pgScorpio) [reply Jamulus.pro]: Wouldn't it be logical to move those somewhere under src too ? (perhaps libs/android?)

Comment (hoffie) [reply Jamulus.pro]: > Wouldn't it be logical to move those somewhere under src too ?

Yes, maybe, but probably not as part of this PR which is titled and branch-named to touch sound files. :)

> (perhaps libs/android?)

It's not a generic lib, rather a platform-specific extension. Also, `libs/` is used for third-party imported code only and it has benefits to keep it that way. As these are not many files, it might be ok to put them directly under src/ and prefix (mac_activity.mm) or suffix (activity_mac.mm) them accordingly. Again... different PR. :)

@ann0see Looks good at first glance, you can run qmake and check the Makefile's clang_format target whether it includes ASIOSDK2 or not (it obviously should not).

Comment (ann0see) [reply Jamulus.pro]: I checked it yesterday and it seems to be ok.

File: Jamulus.pro
Comment (ann0see) [reply Jamulus.pro:198]: Yes that's a good suggestion but probably part of another change/PR (maybe including some other smaller improvements of the file in general).

File: COMPILING.md
Comment (ann0see) [reply COMPILING.md:93]: Copying the contents is more work than just extracting it – at least I think that's the case from CLI/Autobuild.

File: Jamulus.pro
Comment (pgScorpio) [reply Jamulus.pro:198]: As a matter of fact I think moving the sound related files is a good reason to change this too at the same time.

File: COMPILING.md
Comment (pgScorpio) [reply COMPILING.md:93]: The document says "create a folder" and "copy the contents of the archive so that..."
I think copying to an existing (empty) folder decreases the chance on mistakes.

Comment (ann0see) [reply COMPILING.md:93]: Hmm. I now tested what the standard Windows 11 behaviour actually is. Right clicking and starting extract gives a folder in a folder – that's indeed confusing. I'd like not to change the repo but better the description since the deploy_windows script should download the SDK and might need to be modified.

Comment (pgScorpio) [reply COMPILING.md:93]: > since the deploy_windows script should download the SDK and might need to be modified.

I wasn't aware of that. Isn't that against all ASIO rules as discussed in [this](https://github.com/jamulussoftware/jamulus/issues/2417#issuecomment-1068947074) issue)

If it is OK to automatically download it in the the deploy_windows script it could just as well be downloaded as a submodule like oboe.

Comment (hoffie) [reply COMPILING.md:93]: >I wasn't aware of that. Isn't that against all ASIO rules as discussed in https://github.com/jamulussoftware/jamulus/issues/2417#issuecomment-1068947074 issue.

I think "might need to be modified" was related to the build script, not to ASIO SDK files.

> If it is OK to automatically download it in the the deploy_windows script it could just as well be downloaded as a submodule like oboe.

Submodules only work for git repositories. ASIO SDK is only provided as a ZIP download.

Comment (pgScorpio) [reply COMPILING.md:93]: > Submodules only work for git repositories. ASIO SDK is only provided as a ZIP download.

OK not a submodule then, but it still could be downloaded from the make file. Either as a separate make target or as part of the build right ?

**Comment by pgScorpio** [comment]
> move sound API files into src/sound folder

To keep related things together please move soundbase.h and soundbase.cpp to ./Sound too !

**Comment by ann0see** [comment]
@pgScorpio moved the soundbase files too.

Also added a README stub which you can use for documenting your new design + ideas.

File: Jamulus.pro
Comment (ann0see) [reply Jamulus.pro]: Fixed.

File: Jamulus.pro
Comment (ann0see) [reply Jamulus.pro:198]: Yes, but I'd really like to have this PR merged soon (and this can be added in another one)

File: COMPILING.md
Comment (ann0see) [reply COMPILING.md:93]: I mean the deploy_windows.ps1 file already does that?

**Comment by ann0see** [comment]
What's missing here?

**Review by pgScorpio — APPROVED** [review]
Looks good to me

**Review by hoffie — APPROVED** [review]
Looks good to me. Thanks! :)

**Comment by ann0see** [comment]
@pljones os this ok to be merged? @pgScorpio depends on this being merged such that he can continue

**Comment by ann0see** [comment]
@jamulussoftware/maindevelopers any review here? I think we should merge this.

**Comment by ann0see** [comment]
Rebased. I'll open another PR moving the other files soon

**Comment by pljones** [comment]
One thing I noticed when I was doing my various changes... linux/sound.h depends on src/server.h.  It shouldn't - not sure if you want to go fixing this here, though.  (It means moving the high resolution timer from src/server.* to src/util.* - was a simple cut-n-paste.)

**Comment by ann0see** [comment]
I wouldn't want to include it here. This should just move the files without many changes

**Comment by pgScorpio** [comment]
> One thing I noticed when I was doing my various changes... linux/sound.h depends on src/server.h. It shouldn't - not sure if you want to go fixing this here, though. 

Should be another PR I think

> (It means moving the high resolution timer from src/server.* to src/util.* - was a simple cut-n-paste.)

To keep things clean I would create a new .h/.cpp i.e. 'timers'.
I personally don't call a timer a 'util' and in my opinion util is already becoming a receptacle.

**Comment by ann0see** [comment]
@pljones could you please review this PR?

**Comment by ann0see** [comment]
@hoffie @npostavs since you're both familiar with the code could you please review? I assume this is safe to merge.

File: Jamulus.pro
Comment (npostavs) [inline Jamulus.pro:129]: Should be `/libs` or `\\libs`, not `\libs`, I think?

Comment (ann0see) [reply Jamulus.pro:129]: Not sure what the Windows convention is?

Comment (ann0see) [reply Jamulus.pro]: Ok. Changed it. I think `\\` can be used for local paths and network paths. At least it's nothing bad.

File: Jamulus.pro
Comment (npostavs) [inline Jamulus.pro:129]: Ah, you added two more unescaped backslashes.

I double checked it, and it seems that qmake treats `\\x` and `\x` the same (with exception of `x = "`), so I guess you could just leave them unescaped, but it still seems cleaner to have it escaped, IMO.

Comment (ann0see) [reply Jamulus.pro:129]: Ok.

**Comment by ann0see** [comment]
Note to maintainers: Should be squash-merged.

**Comment by ann0see** [comment]
Since @pgScorpio depends on this and I think it's save, I will merge this in the following days if nobody objects

**Comment by ann0see** [comment]
@jamulussoftware/maindevelopers I will merge this today

**Pull request merged**
